llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -1031,6 +1031,27 @@ void State::addInfoForInductions(BasicBlock &BB) {
   WorkList.push_back(FactOrCheck::getConditionFact(
       DTN, CmpInst::ICMP_SLT, PN, B,
       ConditionTy(CmpInst::ICMP_SLE, StartValue, B)));
+
+  assert(!StepOffset.isNegative() && "induction must be increasing");
+  // Try to add condition from header to the unique exit block, if there is one.
+  // When exiting either with EQ or NE, we know that the induction value must be
+  // u<= B, as a different exit may exit earlier.
+  ConditionTy Precond;
+  if (!MonotonicallyIncreasingUnsigned)
+    Precond = {CmpInst::ICMP_ULE, StartValue, B};
+  if (Pred == CmpInst::ICMP_EQ) {
+    BasicBlock *EB = cast<BranchInst>(BB.getTerminator())->getSuccessor(0);
+    if (L->getUniqueExitBlock() == EB) {
+      WorkList.emplace_back(FactOrCheck::getConditionFact(
+          DT.getNode(EB), CmpInst::ICMP_ULE, A, B, Precond));
+    }
+  }
+  if (Pred == CmpInst::ICMP_NE) {
+    BasicBlock *EB = cast<BranchInst>(BB.getTerminator())->getSuccessor(1);
+    if (L->getUniqueExitBlock() == EB)
```

Comment:
I'm not sure I understand the significance of the unique exit block here. Even if the exit is non-unique, wouldn't we be still add this condition fact to the header exit?

Reply:
> I'm not sure I understand the significance of the unique exit block here. Even if the exit is non-unique, wouldn't we be still add this condition fact to the header exit?

It may cause a miscompilation if there is a path from another exit block to the header exit.

Reply:
Why? Isn't the argument here that even if we go through another exit, it can only exit earlier than the header, so the condition still holds? Why would it matter whether the other exit has a direct edge to the header exit vs going there indirectly?

Reply:
It's not needed I think, removed the restriction (for the case where each exit block only has a single predecessor in the loop, the patch isn't needed). Tests added in 798754f6c6fe19d19984ca8370806cbcdf5f8294